File: Refund/Components/Histogram/Histogram.razor.js
const histograms = new Map();

/** Debounce helper with cancel support */
function debounce(fn, wait) {
    let t;
    function debounced(...args) {
        clearTimeout(t);
        t = setTimeout(() => fn.apply(this, args), wait);
    }
    
    // Add cancel method
    debounced.cancel = function() {
        clearTimeout(t);
    };
    
    return debounced;
}

class Histogram {
    constructor(histogramId, rangeId, rangeStartHandleId, rangeEndHandleId, dotNetRef) {
        // DOM references
        this.svg = document.getElementById(histogramId);
        this.id = histogramId;
        this.rangeSelection = document.getElementById(rangeId);
        this.rangeStartHandle = document.getElementById(rangeStartHandleId);
        this.rangeEndHandle = document.getElementById(rangeEndHandleId);
        this.dotNetRef = dotNetRef;

        // Config
        this.binSizes = [];
        this.secondaryBinSizes = [];
        this.minRange = 0;
        this.maxRange = 1;
        this.color = '#0078D4';
        this.secondaryColor = '#8A2BE2';
        this.rangeSelectionEnabled = false;
        this.selectedRangeStart = 0;
        this.selectedRangeEnd = 1;
        this.stepSize = 0.1;
        this.minGap = 1;
        this.disabled = false;
        
        // Debounced method for notifying range changes
        this._debouncedNotifyRangeChanged = debounce(() => {
            this.notifyRangeChanged();
        }, 50);

        // State
        this.isDraggingStart = false;
        this.isDraggingEnd = false;
        this.svgRect = null;
        this.barWidth = 0;

        // Bind event handlers
        this.handleMouseMove = this.handleMouseMove.bind(this);
        this.handleMouseUp = this.handleMouseUp.bind(this);
        this.handleMouseLeave = this.handleMouseLeave.bind(this);
        this.handleResize = debounce(this.handleResize.bind(this), 100);
        
        // Set up event listeners for range handles
        if (this.rangeStartHandle) {
            this.rangeStartHandle.addEventListener('mousedown', (e) => {
                if (this.disabled) return;
                this.isDraggingStart = true;
                this.rangeStartHandle.classList.add('dragging');
                e.preventDefault();
            });
        }
        
        if (this.rangeEndHandle) {
            this.rangeEndHandle.addEventListener('mousedown', (e) => {
                if (this.disabled) return;
                this.isDraggingEnd = true;
                this.rangeEndHandle.classList.add('dragging');
                e.preventDefault();
            });
        }
        
        // Global mouse events
        document.addEventListener('mousemove', this.handleMouseMove);
        document.addEventListener('mouseup', this.handleMouseUp);
        
        // Mouse leave event on the entire histogram container (parent of SVG)
        if (this.svg && this.svg.parentElement) {
            this.svg.parentElement.addEventListener('mouseleave', this.handleMouseLeave);
        }
        
        // Resize observer
        setTimeout(() => {
            this.resizeObserver = new ResizeObserver(() => {
                this.handleResize();
            });
            if (this.svg.parentElement) {
                this.resizeObserver.observe(this.svg.parentElement);
            }
        }, 30);
    }

    update(config) {
        if (!config) return;
        
        // Update configuration
        this.binSizes = config.binSizes || [];
        this.secondaryBinSizes = config.secondaryBinSizes || [];
        this.minRange = config.minRange;
        this.maxRange = config.maxRange;
        this.color = config.color || '#0078D4';
        this.secondaryColor = config.secondaryColor || '#8A2BE2';
        this.rangeSelectionEnabled = config.rangeSelectionEnabled || false;
        this.selectedRangeStart = config.selectedRangeStart;
        this.selectedRangeEnd = config.selectedRangeEnd;
        this.stepSize = config.stepSize || 0.1;
        this.minGap = config.minGap || 1;
        this.disabled = config.disabled || false;
        
        // Render histogram and range selection
        this.renderHistogram();
        this.updateRangeSelection();
    }

    renderHistogram() {
        // Clear existing content
        while (this.svg.firstChild) {
            this.svg.removeChild(this.svg.firstChild);
        }
        
        // Get SVG dimensions
        this.svgRect = this.svg.getBoundingClientRect();
        const width = this.svgRect.width;
        const height = this.svgRect.height;
        
        // Skip if dimensions are too small or no data
        if (width < 10 || height < 10 || this.binSizes.length === 0) return;
        
        // Calculate bar width (no gaps)
        this.barWidth = width / this.binSizes.length;
        
        // Find the maximum bin size across both datasets for scaling
        const maxBinSize = Math.max(
            ...this.binSizes,
            ...(this.secondaryBinSizes.length > 0 ? this.secondaryBinSizes : [0])
        );
        
        // Default to 1 if all bins are empty
        const normFactor = maxBinSize > 0 ? maxBinSize : 1;
        
        // Create a group for primary bars
        const primaryGroup = document.createElementNS('http://www.w3.org/2000/svg', 'g');
        primaryGroup.setAttribute('class', 'histogram-bars-primary');
        this.svg.appendChild(primaryGroup);
        
        // Create primary bars
        for (let i = 0; i < this.binSizes.length; i++) {
            const barHeight = this.binSizes[i] / normFactor * height;
            const x = i * this.barWidth;
            const y = height - barHeight;
            
            const rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
            rect.setAttribute('class', 'histogram-bar primary');
            rect.setAttribute('x', x);
            rect.setAttribute('y', y);
            rect.setAttribute('width', this.barWidth);
            rect.setAttribute('height', barHeight);
            rect.setAttribute('fill', this.color);
            rect.setAttribute('fill-opacity', '0.5'); // Make semi-transparent for overlapping
            rect.setAttribute('data-index', i);
            
            // Add tooltip for bar value range
            const binStart = this.getBinValue(i);
            const binEnd = this.getBinValue(i + 1);
            rect.setAttribute('title', `Primary: ${binStart} - ${binEnd}`);
            
            primaryGroup.appendChild(rect);
        }
        
        // Create secondary bars if they exist
        if (this.secondaryBinSizes.length > 0) {
            // Create a group for secondary bars
            const secondaryGroup = document.createElementNS('http://www.w3.org/2000/svg', 'g');
            secondaryGroup.setAttribute('class', 'histogram-bars-secondary');
            this.svg.appendChild(secondaryGroup);
            
            // Create secondary bars
            for (let i = 0; i < this.secondaryBinSizes.length && i < this.binSizes.length; i++) {
                const barHeight = this.secondaryBinSizes[i] / normFactor * height;
                const x = i * this.barWidth;
                const y = height - barHeight;
                
                const rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
                rect.setAttribute('class', 'histogram-bar secondary');
                rect.setAttribute('x', x);
                rect.setAttribute('y', y);
                rect.setAttribute('width', this.barWidth);
                rect.setAttribute('height', barHeight);
                rect.setAttribute('fill', this.secondaryColor);
                rect.setAttribute('fill-opacity', '0.5'); // Make semi-transparent for overlapping
                rect.setAttribute('data-index', i);
                
                // Add tooltip for bar value range
                const binStart = this.getBinValue(i);
                const binEnd = this.getBinValue(i + 1);
                rect.setAttribute('title', `Secondary: ${binStart} - ${binEnd}`);
                
                secondaryGroup.appendChild(rect);
            }
        }
    }
    
    getBinValue(index) {
        if (this.binSizes.length === 0) return 0;
        
        // Calculate the value at the given bin index
        const range = this.maxRange - this.minRange;
        const binWidth = range / this.binSizes.length;
        return this.minRange + (binWidth * index);
    }
    
    // Get the bin width
    getBinWidth() {
        if (this.binSizes.length <= 1) return 1;
        
        const range = this.maxRange - this.minRange;
        return range / this.binSizes.length;
    }
    
    // Get effective step size, using bin width if step size is 0 or negative
    getEffectiveStepSize() {
        if (this.stepSize <= 0) {
            return this.getBinWidth();
        }
        return this.stepSize;
    }
    
    updateRangeSelection() {
        if (!this.rangeSelectionEnabled || !this.rangeSelection) return;
        
        // Get the range selection area dimensions
        const range = this.maxRange - this.minRange;
        if (range <= 0) return;
        
        // Calculate positions as percentages
        const startPct = (this.selectedRangeStart - this.minRange) / range;
        const endPct = (this.selectedRangeEnd - this.minRange) / range;
        
        // Apply to range background
        this.rangeSelection.style.display = 'block';
        this.rangeSelection.style.left = `${startPct * 100}%`;
        this.rangeSelection.style.width = `${(endPct - startPct) * 100}%`;
        
        // Update handle positions
        if (this.rangeStartHandle) {
            this.rangeStartHandle.style.left = `${startPct * 100}%`;
        }
        
        if (this.rangeEndHandle) {
            this.rangeEndHandle.style.right = `${(1 - endPct) * 100}%`;
        }
    }
    
    handleMouseMove(e) {
        if (this.disabled || (!this.isDraggingStart && !this.isDraggingEnd)) return;
        if (!this.svgRect) this.svgRect = this.svg.getBoundingClientRect();
        
        // Get mouse position relative to SVG
        const relX = (e.clientX - this.svgRect.left) / this.svgRect.width;
        const clampedX = Math.max(0, Math.min(1, relX));
        
        // Convert to value space
        const range = this.maxRange - this.minRange;
        let newValue = this.minRange + (clampedX * range);
        
        // Apply step size (snap to grid)
        const effectiveStepSize = this.getEffectiveStepSize();
        newValue = Math.round(newValue / effectiveStepSize) * effectiveStepSize;
        
        // Handle dragging of different edges
        if (this.isDraggingStart) {
            // Ensure start doesn't exceed end - minGap
            const minGapValue = this.minGap * this.getEffectiveStepSize();
            const maxStart = this.selectedRangeEnd - minGapValue;
            newValue = Math.min(newValue, maxStart);
            
            if (newValue !== this.selectedRangeStart) {
                this.selectedRangeStart = newValue;
                this.updateRangeSelection();
                // Don't notify during drag - only on mouse up or mouse leave
            }
        } else if (this.isDraggingEnd) {
            // Ensure end doesn't go below start + minGap
            const minGapValue = this.minGap * this.getEffectiveStepSize();
            const minEnd = this.selectedRangeStart + minGapValue;
            newValue = Math.max(newValue, minEnd);
            
            if (newValue !== this.selectedRangeEnd) {
                this.selectedRangeEnd = newValue;
                this.updateRangeSelection();
                // Don't notify during drag - only on mouse up or mouse leave
            }
        }
    }
    
    handleMouseUp() {
        let wasDragging = this.isDraggingStart || this.isDraggingEnd;
        
        if (this.isDraggingStart) {
            this.isDraggingStart = false;
            if (this.rangeStartHandle) {
                this.rangeStartHandle.classList.remove('dragging');
            }
        }
        
        if (this.isDraggingEnd) {
            this.isDraggingEnd = false;
            if (this.rangeEndHandle) {
                this.rangeEndHandle.classList.remove('dragging');
            }
        }
        
        // If we were dragging, send a final non-debounced update
        if (wasDragging) {
            // Cancel any pending debounced updates
            if (this._debouncedNotifyRangeChanged.cancel) {
                this._debouncedNotifyRangeChanged.cancel();
            }
            
            // Send the final update immediately
            this.notifyRangeChanged();
        }
    }
    
    handleMouseLeave() {
        // Stop dragging when mouse leaves the histogram component
        let wasDragging = this.isDraggingStart || this.isDraggingEnd;
        
        if (this.isDraggingStart) {
            this.isDraggingStart = false;
            if (this.rangeStartHandle) {
                this.rangeStartHandle.classList.remove('dragging');
            }
        }
        
        if (this.isDraggingEnd) {
            this.isDraggingEnd = false;
            if (this.rangeEndHandle) {
                this.rangeEndHandle.classList.remove('dragging');
            }
        }
        
        // If we were dragging, send a final update
        if (wasDragging) {
            // Cancel any pending debounced updates
            if (this._debouncedNotifyRangeChanged.cancel) {
                this._debouncedNotifyRangeChanged.cancel();
            }
            
            // Send the final update immediately
            this.notifyRangeChanged();
        }
    }
    
    handleResize() {
        this.svgRect = null; // Force recalculation of dimensions
        this.renderHistogram();
        this.updateRangeSelection();
    }
    
    notifyRangeChanged() {
        if (this.dotNetRef) {
            this.dotNetRef.invokeMethodAsync('OnRangeChanged', 
                this.selectedRangeStart, 
                this.selectedRangeEnd);
        }
    }
    
    dispose() {
        // Remove event listeners
        document.removeEventListener('mousemove', this.handleMouseMove);
        document.removeEventListener('mouseup', this.handleMouseUp);
        
        // Remove mouse leave event listener
        if (this.svg && this.svg.parentElement) {
            this.svg.parentElement.removeEventListener('mouseleave', this.handleMouseLeave);
        }
        
        // Disconnect resize observer
        if (this.resizeObserver) {
            this.resizeObserver.disconnect();
        }
    }
}

/** Blazor export: initialize a new histogram */
export function initializeHistogram(
    histogramId,
    rangeId,
    rangeStartHandleId,
    rangeEndHandleId,
    dotNetRef,
    config
) {
    const histogram = new Histogram(
        histogramId,
        rangeId,
        rangeStartHandleId,
        rangeEndHandleId,
        dotNetRef
    );
    histograms.set(histogramId, histogram);
    histogram.update(config);
}

/** Blazor export: update an existing histogram */
export function updateHistogram(histogramId, config) {
    const histogram = histograms.get(histogramId);
    if (histogram) histogram.update(config);
}

/** Blazor export: dispose a histogram */
export function disposeHistogram(histogramId) {
    const histogram = histograms.get(histogramId);
    if (histogram) {
        histogram.dispose();
        histograms.delete(histogramId);
    }
}

File: Refund/Components/Histogram/Histogram.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.Extensions.Logging;
using Microsoft.JSInterop;
using System.Globalization;

namespace Refund.Components.Histogram;

public partial class Histogram : IAsyncDisposable
{
    #region Parameters

    /// <summary>
    /// The bin sizes for the primary histogram. Each value represents the height of a bin.
    /// </summary>
    [Parameter] public decimal[] BinSizes { get; set; } = Array.Empty<decimal>();
    private decimal[] _previousBinSizes = Array.Empty<decimal>();

    /// <summary>
    /// The bin sizes for the secondary histogram. Each value represents the height of a bin.
    /// If provided, will render as an overlapping histogram with the primary one.
    /// Must have the same length as BinSizes to render properly.
    /// </summary>
    [Parameter] public decimal[] SecondaryBinSizes { get; set; } = Array.Empty<decimal>();
    private decimal[] _previousSecondaryBinSizes = Array.Empty<decimal>();

    /// <summary>
    /// The color for the secondary histogram bars.
    /// </summary>
    [Parameter] public string SecondaryColor { get; set; } = "#8A2BE2"; // Default to BlueViolet

    /// <summary>
    /// The minimum value represented in the histogram (left edge of first bin).
    /// </summary>
    [Parameter] public decimal MinRange { get; set; } = 0;

    /// <summary>
    /// The maximum value represented in the histogram (right edge of last bin).
    /// </summary>
    [Parameter] public decimal MaxRange { get; set; } = 1;

    /// <summary>
    /// The fill color for the histogram bars.
    /// </summary>
    [Parameter] public string Color { get; set; } = "#0078D4";

    /// <summary>
    /// Whether to enable range selection on the histogram.
    /// </summary>
    [Parameter] public bool RangeSelectionEnabled { get; set; } = false;
    
    /// <summary>
    /// Whether the component is disabled (disables all interaction).
    /// </summary>
    [Parameter] public bool IsDisabled { get; set; } = false;

    /// <summary>
    /// The start value of the selected range.
    /// </summary>
    [Parameter] public decimal SelectedRangeStart { get; set; } = 0;

    /// <summary>
    /// The end value of the selected range.
    /// </summary>
    [Parameter] public decimal SelectedRangeEnd { get; set; } = 1;
    
    /// <summary>
    /// Event fired when the selected range changes.
    /// </summary>
    [Parameter] public EventCallback<(decimal Start, decimal End)> SelectedRangeChanged { get; set; }

    /// <summary>
    /// The step size for range adjustments.
    /// </summary>
    [Parameter] public decimal StepSize { get; set; } = 0.1m;

    /// <summary>
    /// The minimum gap (in steps) between the start and end of the range.
    /// </summary>
    [Parameter] public int MinGap { get; set; } = 1;
    
    /// <summary>
    /// Optional CSS class for the container.
    /// </summary>
    [Parameter] public string? Class { get; set; }

    /// <summary>
    /// Optional inline style for the container.
    /// </summary>
    [Parameter] public string? Style { get; set; }

    /// <summary>
    /// Catch-all for additional attributes.
    /// </summary>
    [Parameter(CaptureUnmatchedValues = true)] public Dictionary<string, object>? AdditionalAttributes { get; set; }

    #endregion

    #region Injected Services

    [Inject] private IJSRuntime JsRuntime { get; set; }
    [Inject] private ILogger<Histogram> Logger { get; set; } = default!;

    #endregion

    #region Internal State

    // Unique element IDs
    private readonly string HistogramId = $"histogram-{Guid.NewGuid():N}";
    private readonly string RangeId = $"histogram-range-{Guid.NewGuid():N}";
    private readonly string RangeStartHandleId = $"histogram-range-start-{Guid.NewGuid():N}";
    private readonly string RangeEndHandleId = $"histogram-range-end-{Guid.NewGuid():N}";

    // JS module reference
    private IJSObjectReference? _module;
    private DotNetObjectReference<Histogram>? _dotNetRef;

    private string MinRangeFormatted => FormatValue(MinRange);
    private string MaxRangeFormatted => FormatValue(MaxRange);
    
    #endregion

    #region Lifecycle

    protected override async Task OnInitializedAsync()
    {
        // Create .NET reference for JS callbacks
        _dotNetRef = DotNetObjectReference.Create(this);
        await base.OnInitializedAsync();
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            // Load JS module
            _module = await JsRuntime.InvokeAsync<IJSObjectReference>(
                "import", "./_content/Refund/Components/Histogram/Histogram.razor.js");

            // Initialize histogram
            await InitializeHistogramAsync();
        }
        
        await base.OnAfterRenderAsync(firstRender);
    }

    protected override async Task OnParametersSetAsync()
    {
        // Check if bin sizes changed
        bool binsChanged = false;
        bool secondaryBinsChanged = false;
        
        if (_previousBinSizes.Length != BinSizes.Length)
        {
            binsChanged = true;
        }
        else
        {
            for (int i = 0; i < BinSizes.Length; i++)
            {
                if (BinSizes[i] != _previousBinSizes[i])
                {
                    binsChanged = true;
                    break;
                }
            }
        }

        // Check if secondary bin sizes changed
        if (_previousSecondaryBinSizes.Length != SecondaryBinSizes.Length)
        {
            secondaryBinsChanged = true;
        }
        else
        {
            for (int i = 0; i < SecondaryBinSizes.Length; i++)
            {
                if (SecondaryBinSizes[i] != _previousSecondaryBinSizes[i])
                {
                    secondaryBinsChanged = true;
                    break;
                }
            }
        }

        // Update previous values for next comparison
        _previousBinSizes = BinSizes.ToArray();
        _previousSecondaryBinSizes = SecondaryBinSizes.ToArray();

        // Update JS if needed
        if (binsChanged || secondaryBinsChanged || MinRange != _previousMinRange || MaxRange != _previousMaxRange || 
            Color != _previousColor || SecondaryColor != _previousSecondaryColor || SelectedRangeStart != _previousRangeStart || 
            SelectedRangeEnd != _previousRangeEnd)
        {
            await UpdateHistogramAsync();
            
            _previousMinRange = MinRange;
            _previousMaxRange = MaxRange;
            _previousColor = Color;
            _previousSecondaryColor = SecondaryColor;
            _previousRangeStart = SelectedRangeStart;
            _previousRangeEnd = SelectedRangeEnd;
        }

        await base.OnParametersSetAsync();
    }

    public async ValueTask DisposeAsync()
    {
        try
        {
            if (_module != null)
            {
                await _module.InvokeVoidAsync("disposeHistogram", HistogramId);
                await _module.DisposeAsync();
            }

            _dotNetRef?.Dispose();
        }
        catch
        {
            // ignored
        }
    }

    #endregion

    #region JS Interop

    private async Task InitializeHistogramAsync()
    {
        if (_module == null) return;
        
        try
        {
            await _module.InvokeVoidAsync("initializeHistogram",
                HistogramId,
                RangeId,
                RangeStartHandleId,
                RangeEndHandleId,
                _dotNetRef,
                GetConfigData());
        }
        catch (Exception ex)
        {
            Logger.LogError(ex, "Error initializing Histogram with ID {HistogramId}", HistogramId);
        }
    }

    private async Task UpdateHistogramAsync()
    {
        if (_module == null) return;
        
        try
        {
            await _module.InvokeVoidAsync("updateHistogram",
                HistogramId,
                GetConfigData());
        }
        catch (Exception ex)
        {
            Logger.LogError(ex, "Error updating Histogram with ID {HistogramId}", HistogramId);
        }
    }

    private object GetConfigData()
    {
        return new
        {
            binSizes = BinSizes,
            secondaryBinSizes = SecondaryBinSizes,
            minRange = MinRange,
            maxRange = MaxRange,
            color = Color,
            secondaryColor = SecondaryColor,
            rangeSelectionEnabled = RangeSelectionEnabled,
            selectedRangeStart = SelectedRangeStart,
            selectedRangeEnd = SelectedRangeEnd,
            stepSize = StepSize,
            minGap = MinGap,
            disabled = IsDisabled
        };
    }

    #endregion

    #region JS -> .NET Callbacks

    [JSInvokable]
    public async Task OnRangeChanged(decimal start, decimal end)
    {
        if (start == SelectedRangeStart && end == SelectedRangeEnd)
        {
            return;
        }
        
        // Update the range values
        SelectedRangeStart = start;
        SelectedRangeEnd = end;
        
        // Update UI and notify parent component immediately
        // JavaScript already handles debouncing - this only gets called on mouse up or mouse leave
        await InvokeAsync(StateHasChanged);
        await SelectedRangeChanged.InvokeAsync((start, end));
    }

    #endregion

    #region Helpers

    // Value change handlers for number inputs
    private async Task OnRangeStartChanged(decimal value)
    {
        // Enforce min gap and constraints
        value = Math.Max(MinRange, value);
        value = Math.Min(SelectedRangeEnd - MinGapInValues, value);
        
        // Apply step size
        decimal stepSize = GetEffectiveStepSize();
        if (stepSize > 0)
        {
            value = Math.Round(value / stepSize) * stepSize;
        }
        
        if (value != SelectedRangeStart)
        {
            SelectedRangeStart = value;
            await UpdateHistogramAsync();
            await SelectedRangeChanged.InvokeAsync((SelectedRangeStart, SelectedRangeEnd));
        }
    }
    
    private async Task OnRangeEndChanged(decimal value)
    {
        // Enforce min gap and constraints
        value = Math.Min(MaxRange, value);
        value = Math.Max(SelectedRangeStart + MinGapInValues, value);
        
        // Apply step size
        decimal stepSize = GetEffectiveStepSize();
        if (stepSize > 0)
        {
            value = Math.Round(value / stepSize) * stepSize;
        }
        
        if (value != SelectedRangeEnd)
        {
            SelectedRangeEnd = value;
            await UpdateHistogramAsync();
            await SelectedRangeChanged.InvokeAsync((SelectedRangeStart, SelectedRangeEnd));
        }
    }
    
    // Calculate minimum gap in terms of values rather than steps
    private decimal MinGapInValues => MinGap * GetEffectiveStepSize();
    
    // Get effective step size, making it relative to the number of bins
    private decimal GetEffectiveStepSize()
    {
        // If step size is 0 or negative, default to 1 bin width
        if (StepSize <= 0)
        {
            return GetBinWidth();
        }
        
        return StepSize;
    }
    
    // Calculate the width of a single bin
    private decimal GetBinWidth()
    {
        if (BinSizes.Length <= 1)
        {
            return 1;
        }
        
        return (MaxRange - MinRange) / BinSizes.Length;
    }

    private string FormatValue(decimal value)
    {
        // Format values based on their magnitude
        if (Math.Abs(value) < 0.01m)
        {
            return value.ToString("0.####", CultureInfo.InvariantCulture);
        }
        else if (Math.Abs(value) < 1m)
        {
            return value.ToString("0.##", CultureInfo.InvariantCulture);
        }
        else if (Math.Abs(value) < 10m)
        {
            return value.ToString("0.#", CultureInfo.InvariantCulture);
        }
        else
        {
            return value.ToString("0", CultureInfo.InvariantCulture);
        }
    }

    // Field tracking for parameter changes
    private decimal _previousMinRange;
    private decimal _previousMaxRange;
    private string _previousColor = "#0078D4";
    private string _previousSecondaryColor = "#8A2BE2";
    private decimal _previousRangeStart;
    private decimal _previousRangeEnd;

    #endregion
}